pkg/search/options.go

```diff
@@ -313,16 +309,78 @@ var (
 	TestChild2Val            = newFieldLabel("Test Child2 Val")
 	TestParent3ID            = newFieldLabel("Test Parent3 ID")
 	TestParent3Val           = newFieldLabel("Test Parent3 Val")
+
+	// Derived test fields
+	TestGrandparentCount = newDerivedFieldLabel("Test Grandparent Count", TestGrandparentID, CountDerivationType)
+	TestParent1Count     = newDerivedFieldLabel("Test Parent1 Count", TestParent1ID, CountDerivationType)
+	TestChild1Count      = newDerivedFieldLabel("Test Child1 Count", TestChild1ID, CountDerivationType)
 )
 
-func newFieldLabel(s string) FieldLabel {
-	if added := FieldLabelSet.Add(s); !added {
+func init() {
+	derivationsByField = make(map[string]map[string]DerivationType)
+	for k, metadata := range allFieldLabels {
+		if metadata != nil {
+			derivedFromLower := strings.ToLower(string(metadata.DerivedFrom))
+			subMap, exists := derivationsByField[derivedFromLower]
+			if !exists {
+				subMap = make(map[string]DerivationType)
+				derivationsByField[derivedFromLower] = subMap
+			}
+			subMap[k] = metadata.DerivationType
+		}
+	}
+}
+
+var (
+	allFieldLabels     = make(map[string]*DerivedFieldLabelMetadata)
+	derivationsByField map[string]map[string]DerivationType
+)
+
+// IsValidFieldLabel returns whether this is a known, valid field label.
+func IsValidFieldLabel(s string) bool {
+	_, ok := allFieldLabels[strings.ToLower(s)]
+	return ok
+}
+
+// GetFieldsDerivedFrom gets the fields derived from the given search field.
+func GetFieldsDerivedFrom(s string) map[string]DerivationType {
+	return derivationsByField[strings.ToLower(s)]
+}
+
+func newFieldLabelWithMetadata(s string, metadata *DerivedFieldLabelMetadata) FieldLabel {
+	lowerS := strings.ToLower(s)
```

Comment:
for some reason lowerS is strange to me. just set 
`s = strings.ToLower(s)`

Reply:
I can't, unfortunately, because I want to return `FieldLabel(s)` below.